common/src/main/java/io/pravega/common/concurrent/AsyncSemaphore.java

```diff
@@ -0,0 +1,220 @@
+     * become availble. There is no prioritization of queued tasks - they are triggered in the order in which they
+     * are queued up.
+     *
+     * A task will allocate the requested credits when it is triggered. If the task fails (synchronously or asynchronously),
+     * then the requested credits are automatically released back into the pool. If the task succeeds, the credits will
+     * remain.
+     *
+     * @param task    A {@link Supplier} that, when invoked, will execute the task.
+     * @param credits The number of credits this task requires.
+     * @param <T>     Return type.
+     * @return A CompletableFuture that, when completed, will contain the result of the executed task. If the task failed
+     * or was rejected (i.e., due to {@link AsyncSemaphore} closing), it will be failed with the appropriate exception.
+     */
+    public <T> CompletableFuture<T> run(@NonNull Supplier<CompletableFuture<T>> task, int credits) {
+        Preconditions.checkArgument(credits >= 0 && credits <= this.totalCredits,
+                "credits must be a non-negative number smaller than or equal to %s.", this.totalCredits);
+
+        DelayedTask<T> dt;
+        synchronized (this.queue) {
+            Exceptions.checkNotClosed(this.closed, this);
+            if (canExecute(credits)) {
+                dt = null;
+                this.usedCredits += credits;
+            } else {
+                // Insufficient credits; need to queue up and execute when more becomes available.
+                dt = new DelayedTask<>(credits, task);
+                this.queue.addLast(dt);
+            }
+
+        }
+        if (dt == null) {
+            // We have more credits than what this task requires. Execute now without queuing.
+            return execute(task, credits);
+        } else {
+            // This wil be completed when its associated task is executed.
+            return dt.result;
+        }
+    }
+
+    /**
+     * Releases a number of credits back into the pool.
+     *
+     * @param credits The number of credits to release.
+     */
+    public void release(int credits) {
+        Preconditions.checkArgument(credits >= 0, "credits must be a non-negative number.");
+        synchronized (this.queue) {
+            Exceptions.checkNotClosed(this.closed, this);
+            this.usedCredits = Math.max(0, this.usedCredits - credits);
+        }
+
+        ArrayList<DelayedTask<?>> toExecute = new ArrayList<>();
+        synchronized (this.queue) {
+            while (!this.queue.isEmpty() && canExecute(this.queue.peekFirst().credits)) {
+                DelayedTask<?> qi = this.queue.removeFirst();
+                this.usedCredits += qi.credits;
+                toExecute.add(qi);
+            }
+        }
+
+        toExecute.forEach(this::execute);
+    }
+
+    private <T> void execute(DelayedTask<T> qi) {
+        execute(qi.runTask, qi.credits)
+                .whenComplete((r, ex) -> {
```

Comment:
`execute(Supplier, int)` is used in two places. One is here, and one is in `run`, where it is invoked immediately if there are sufficient credits available. There is no `PendingTask` in `run` for this invocation, so I can't pass that as an argument. Whichever way I do it, there will still need to be a `whenComplete` so that I can pass on the result to `PendingTask.result` (which is what the external caller is waiting for).